mixer/adapter/kubernetesenv/kubernetesenv.go

```diff
@@ -123,30 +130,60 @@ func (b *builder) Build(ctx context.Context, env adapter.Env) (adapter.Handler,
 	// provide this basic functionality before.
 	b.Lock()
 	defer b.Unlock()
-	controller, found := b.controllers[path]
+	_, found := b.controllers[path]
 	if !found {
 		clientset, err := b.newClientFn(path, env)
 		if err != nil {
 			return nil, fmt.Errorf("could not build kubernetes client: %v", err)
 		}
-		controller = newCacheController(clientset, refresh, env)
-		env.ScheduleDaemon(func() { controller.Run(stopChan) })
-		// ensure that any request is only handled after
-		// a sync has occurred
-		env.Logger().Infof("Waiting for kubernetes cache sync...")
-		if success := cache.WaitForCacheSync(stopChan, controller.HasSynced); !success {
-			stopChan <- struct{}{}
-			return nil, errors.New("cache sync failure")
+		controller, err = getNewCacheController(b, clientset, env)
+		if err != nil {
+			return nil, fmt.Errorf("could not create new cache controller: %v", err)
 		}
-		env.Logger().Infof("Cache sync successful.")
+		controllers[path] = controller
 		b.controllers[path] = controller
+	} else {
+		for clusterID := range b.controllers {
+			controllers[clusterID] = b.controllers[clusterID]
+		}
 	}
 
-	return &handler{
+	kubeHandler := handler{
 		env:      env,
-		k8sCache: controller,
+		k8sCache: controllers,
 		params:   paramsProto,
-	}, nil
+	}
+
+	b.kubeHandler = &kubeHandler
+
+	if !found {
+		if err := initMultiClusterSecretController(b, path, env); err != nil {
+			return nil, fmt.Errorf("could not create remote controllers: %v", err)
+		}
+	}
+
+	return &kubeHandler, nil
+}
+
+func getNewCacheController(b *builder, clientset k8s.Interface,
+	env adapter.Env) (cacheController, error) {
```

Comment:
super nit: i'd just move this onto a single line for readability.

Reply:
Done. You've mentioned this before and I try my best to remember that istio has a 160 column limit.
Just doesn't seem right to me. https://twitter.com/i/status/999652792707858433

Reply:
lol. now i need a gif for trying to decide if a line is code or part of the signature :)